Add R documentation

Cimbali opened the pull request:
Contains the R base + recommended package help pages converted to HTML.

Equivalent to the fullrefman.pdf generated from source, which is also called « The R Reference Index » on https://cran.r-project.org/manuals.html <em>plus the intro and data import/export manuals</em>.

~~Currently does not include reference manuals and miscellanea (FAQ, etc.), you can include them easily b y changing the class-level variables on the entries filter (`@@include_manual`, `@@include_misc`).~~

Script building the documentation below (I’m also attaching a [zipball of the built documentation](https://github.com/freeCodeCamp/devdocs/files/6580874/r.zip)).

Note that the license file and link indicate GPLv2, however I simply copied the line from the R documentation template that says « under the GPL » (https://svn.r-project.org/R/trunk/doc/manual/refman.top), without specifying if it’s v2, v2+ (or maybe even v1+).

```bash
DEVDOCSROOT=$HOME/projects/devdocs/docs/R
RLATEST=https://cran.r-project.org/src/base/R-latest.tar.gz

RSOURCEDIR=${TMPDIR:-/tmp}/R/latest
RBUILDDIR=${TMPDIR:-/tmp}/R/build
mkdir -p "$RSOURCEDIR" "$RBUILDDIR" "$DEVDOCSROOT"

# Download, configure, and build with static HTML pages
curl "$RLATEST" | tar -C "$RSOURCEDIR" -xzf - --strip-components=1
(cd "$RBUILDDIR" && "$RSOURCEDIR/configure" --enable-prebuilt-html --with-recommended-packages --disable-byte-compiled-packages --disable-shared --disable-java)
make _R_HELP_LINKS_TO_TOPICS_=FALSE -C "$RBUILDDIR"

# Export all html documentation built − global, and per-package
cp -r "$RBUILDDIR/doc" "$DEVDOCSROOT/"
ls -d "$RBUILDDIR"/library/*/html | while read orig; do
    dest="$DEVDOCSROOT${orig#$RBUILDDIR}"
    mkdir -p "$dest" && cp -r "$orig"/* "$dest/"
done
```


- [x] Tested the scraper on a local copy of DevDocs
- [x] Ensured that the docs are styled similarly to other docs on DevDocs
<!-- If the docs don’t have an icon, delete the next four items: -->
- [x] Added these files to the <code>public/icons/*your_scraper_name*/</code> directory:
  - [x] `16.png`: a 16×16 pixel icon for the doc
  - [x] `16@2x.png`: a 32×32 pixel icon for the doc
  - [x] `SOURCE`: A text file containing the URL to the page the image can be found on or the URL of the original image itself

simon04 commented on assets/javascripts/templates/pages/about_tmpl.coffee:
```diff
@@ -676,6 +676,11 @@ credits = [
     'https://doc.qt.io/qt-5/licensing.html'
+  ], [
+    'R',
+    '1999--2012 R Foundation for Statistical Computing',
```
```suggestion
    '1999–2012 R Foundation for Statistical Computing',
```

simon04 commented on lib/docs/scrapers/r.rb:
```diff
@@ -0,0 +1,31 @@
+    options[:skip_links] = false
+
+    options[:attribution] = <<-HTML
+      Copyright (&copy;) 1999--2012 R Foundation for Statistical Computing.<br>
```
```suggestion
      Copyright (&copy;) 1999–2012 R Foundation for Statistical Computing.<br>
```

Cimbali commented:
By the way we should probably either remove or enable the dead code (manuals and misc pages) but I’m not sure which way to go on this.

Some of it is interesting (R introduction mostly ? probably data import/export too) and some of it seems too much (writing extensions and internals seem too advanced) and some clearly overboard (Installation, language definition). However it would add a lot of noise to the docs.

Any advice welcome.

Cimbali commented:
I believe the R scraper is a little more satisfactory now. It’s including the 2 basic manuals and all the default package index. HTML seems clean enough to me.

I’d like to do some URL rewrite on a few links that point to existing entries or packages, but where the target file does not exist.
There’s 2 things basically:
- links to `/library/${package}/doc/index.html` should point to the root of the package, i.e. `/r-${package}/`
- links to `/library/${package}/html/${missingpage}.html` should point to the entry (which does exist), i.e. `/library/${package}/html/${missingpage}` (transformed as a proper slug)

To make this tricky, there are also links to packages that are not included in the standard install (so there not finding the link target is normal).

I’m not sure how to go about it. It seems to me that at the moment of cleaning a page’s HTML, the full list of entries is not yet available, meaning I can not check for a package’s existence.

It’s only a dozen dead links like that, so they could be manually handled too, but that doesn’t seem future-proof.

MasterEnoc commented:
Would this scraper work using the prebuilt HTML files found in the [debian package 'r-doc-html'](https://packages.debian.org/search?keywords=r-doc-html&searchon=names&suite=all&section=all)? Devdocs already has a documentation whose HTMLs files are obtained in this way, check https://github.com/freeCodeCamp/devdocs/blob/main/docs/file-scrapers.md#openjdk.

Cimbali commented:
As you can see from [the list of files in the package](https://packages.debian.org/jessie/all/r-doc-html/filelist), the Debian package only contains the few manuals (R intro, R data, R admin, etc), and not the html pages of all base packages, which are contained only in `fullrefman.pdf`, which is only in the [pdf doc](https://packages.debian.org/search?keywords=r-doc-pdf). Same as from the R website.

simon04 commented:
Some special rules for redirects are specified in app.rb. You might be able to accomplish your needs this way? https://github.com/freeCodeCamp/devdocs/blob/fb7db1e25348af4edc38dfa2b9f90eb03da3e508/lib/app.rb#L369-L406

Cimbali commented:
Thanks for the suggestion @simon04, however that didn’t quite work out. I found another way, by generating the links to rewrite at parsing time (from the entries filter).

 <details>
<summary>I tried the following in lib/app</summary>

```diff
diff --git a/lib/app.rb b/lib/app.rb
index e046dc4b..0d85c028 100644
--- a/lib/app.rb
+++ b/lib/app.rb
@@ -403,6 +403,18 @@ class App < Sinatra::Application
       end
     end
 
+    if rest && doc == 'r' && rest.start_with?('/library/') && !File.exist?(File.join(App.docs_path, doc, "#{rest}.html"))
+      begin
+        base, package, _html, keyword = rest.delete_prefix('/').split('/')
+      rescue
+        next
+      end
+      entry = JSON.parse(File.read(File.join(App.docs_path, doc, 'index.json')))['entries'].find do |obj|
+        obj['type'].downcase == package and obj['name'] == keyword
+      end
+      return redirect "/r/#{entry['path']}" unless entry.nil?
+    end
+
     return 404 unless @doc = find_doc(doc)
 
     if rest.nil?
```

This works when entering the URL of the page, i.e. navigating to `http://localhost:9292/r/library/mass/html/cov.mve` correctly redirects to `http://localhost:9292/r/library/mass/html/cov.rob`. However, I still get a 404 when clicking a link from another page − e.g. the `cov.mve` link in the « See Also » section of `http://localhost:9292/r/library/cluster/html/ellipsoidhull`.
</details>


I think that’s all there is to the R docs:

- I’ve simplified the script to generate the docs (updated in the post above): [r.zip](https://github.com/freeCodeCamp/devdocs/files/6580874/r.zip)
- scraper / html cleaner updated accordingly
- all remaining issues (such as dead links discussed) are now resolved

simon04 reviewed (approved):
Great, thank you!

MasterEnoc commented:
@Cimbali Could you add an entry in https://github.com/freeCodeCamp/devdocs/blob/main/docs/file-scrapers.md with the steps to scrap the R documentation?

Dwalinnewyork commented:
DEVDOCSROOT=$HOME/projects/devdocs/docs/R
RLATEST=https://cran.r-project.org/src/base/R-latest.tar.gz

RSOURCEDIR=${TMPDIR:-/tmp}/R/latest
RBUILDDIR=${TMPDIR:-/tmp}/R/build
mkdir -p "$RSOURCEDIR" "$RBUILDDIR" "$DEVDOCSROOT"

# Download, configure, and build with static HTML pages
curl "$RLATEST" | tar -C "$RSOURCEDIR" -xzf - --strip-components=1
(cd "$RBUILDDIR" && "$RSOURCEDIR/configure" --enable-prebuilt-html --with-recommended-packages --disable-byte-compiled-packages --disable-shared --disable-java)
make _R_HELP_LINKS_TO_TOPICS_=FALSE -C "$RBUILDDIR"

# Export all html documentation built − global, and per-package
cp -r "$RBUILDDIR/doc" "$DEVDOCSROOT/"
ls -d "$RBUILDDIR"/library/*/html | while read orig; do
    dest="$DEVDOCSROOT${orig#$RBUILDDIR}"
    mkdir -p "$dest" && cp -r "$orig"/* "$dest/"
done

Dwalinnewyork commented:
DEVDOCSROOT=$HOME/projects/devdocs/docs/R
RLATEST=https://cran.r-project.org/src/base/R-latest.tar.gz

RSOURCEDIR=${TMPDIR:-/tmp}/R/latest
RBUILDDIR=${TMPDIR:-/tmp}/R/build
mkdir -p "$RSOURCEDIR" "$RBUILDDIR" "$DEVDOCSROOT"

# Download, configure, and build with static HTML pages
curl "$RLATEST" | tar -C "$RSOURCEDIR" -xzf - --strip-components=1
(cd "$RBUILDDIR" && "$RSOURCEDIR/configure" --enable-prebuilt-html --with-recommended-packages --disable-byte-compiled-packages --disable-shared --disable-java)
make _R_HELP_LINKS_TO_TOPICS_=FALSE -C "$RBUILDDIR"

# Export all html documentation built − global, and per-package
cp -r "$RBUILDDIR/doc" "$DEVDOCSROOT/"
ls -d "$RBUILDDIR"/library/*/html | while read orig; do
    dest="$DEVDOCSROOT${orig#$RBUILDDIR}"
    mkdir -p "$dest" && cp -r "$orig"/* "$dest/"
done

Dwalinnewyork commented:
DEVDOCSROOT=$HOME/projects/devdocs/docs/R
RLATEST=https://cran.r-project.org/src/base/R-latest.tar.gz

RSOURCEDIR=${TMPDIR:-/tmp}/R/latest
RBUILDDIR=${TMPDIR:-/tmp}/R/build
mkdir -p "$RSOURCEDIR" "$RBUILDDIR" "$DEVDOCSROOT"

# Download, configure, and build with static HTML pages
curl "$RLATEST" | tar -C "$RSOURCEDIR" -xzf - --strip-components=1
(cd "$RBUILDDIR" && "$RSOURCEDIR/configure" --enable-prebuilt-html --with-recommended-packages --disable-byte-compiled-packages --disable-shared --disable-java)
make _R_HELP_LINKS_TO_TOPICS_=FALSE -C "$RBUILDDIR"

# Export all html documentation built − global, and per-package
cp -r "$RBUILDDIR/doc" "$DEVDOCSROOT/"
ls -d "$RBUILDDIR"/library/*/html | while read orig; do
    dest="$DEVDOCSROOT${orig#$RBUILDDIR}"
    mkdir -p "$dest" && cp -r "$orig"/* "$dest/"
done
